eth/db/chain.py

```diff
@@ -151,6 +152,12 @@ def _persist_block(
             else:
                 tx_hashes = cls._get_block_transaction_hashes(db, header)
 
+            tx_root_hash, tx_kv_nodes = make_trie_root_and_nodes(block.transactions)
+            cls._persist_trie_data_dict(db, tx_kv_nodes)
+
```

Comment:
@carver I'm hitting a little wall here. I was able to write a test that proofs tx aren't saved from `persist_block` unless we include these lines :point_up: 

But I wonder how would I save the receipts here. If I understand correctly, the receipts are a byproduct of block execution which means we would not have them if we don't execute them. My best guess is that we should add another (optional) parameter to `persist_block` that would take receipts (provided by peers during backfill), verifies them, and then stores them accordingly. Is that roughly what we want?

Reply:
Hm, well the weird thing is that now the transactions are persisted twice on a `Chain.import_block()`, which calls both `import_block()` and `persist_block()`.

So I am pretty tempted to shake up the API a bit more here and say:
- `VM.import_block()` stops being responsible for saving *anything* to the database: it will be read-only, and return anything that needs to be saved (I think it returns almost enough, except for receipts)
- `ChainDB.persist_block()` takes a new *non-optional* receipts argument, and is solely responsible for saving transactions/receipts/etc

Then we could get rid of this hacky bit, too:
https://github.com/ethereum/trinity/blob/32f523128780fc7e24e1582b4bd8d87813396763/trinity/sync/full/chain.py#L885-L886

Because we would have a more formal API for persisting block/transaction/receipt.

Reply:
Yes, the double persist is another sub optimal thing. I wasn't exactly sure if there might be a strong reason why we would need to persist directly from `import_block` but I agree we should try to get rid of that and tighten the API :+1:

Reply:
It occurs to me that it might be difficult to delay persisting the state database changes, but maybe possible. If that turns out to be too big a rabbit-hole, then we can just add an issue and move on...

Reply:
I'm looking into this but it it may be more complicated than I thought. E.g. another issue is, as far as I understand we can not delay the calculation of  `receipt_root` and the `transaction_root` until `persist_block`. But if we already calculate that early, then the only way to avoid recalculating it later in `persist_block` would be to cache and pass around all the created nodes which doesn't seem viable. But duplicating the work for every block doesn't sound great either.

I think it just comes down to the fact that there are two fundamental different ways that may lead to the persistence of blocks:

1. We have executed a block and want to persist it
2. We have downloaded a block and have **no desire to execute** it but want to persist it.

Only in the second case would it be reasonable to do the extra work of calculating the `receipt_root` and `transaction_root` as a pre-step to persistence. In the other case (1.) we can not delay this step because even if we do not aim to persist the block just yet, we sure want a full in-memory representation of it **prior to persisting** which just happens to depend on having `receipt_root` and `transaction_root` already calculated.

So, what I'm thinking is that maybe we simple want a new `ChainDB` API `persist_unexecuted_block`:

```python
    def persist_unexecuted_block(self,
                      block: BlockAPI,
                      receipts: Tuple[ReceiptAPI, ...],
                      genesis_parent_hash: Hash32 = GENESIS_PARENT_HASH
                      ) -> Tuple[Tuple[Hash32, ...], Tuple[Hash32, ...]]:
```

This isn't much different than adding an *optional*  `receipts` argument to `persist_block` but the separation allows to:
  - only calculate and store `tx_nodes` when we know we have to (it would also be a bit awkward to conditionally do that task based on the fact if `receipts` is `None` or not.
 - only calculate and store `receipt_nodes` when we know we have to
 - only do the verification when we know we have to
 - have a tighter, more meaningful API

Reply:
Oh btw, I just noticed that we even had it documented that `persist_block` is based on the assumption that transactions were already persisted seperatedly.

https://github.com/ethereum/py-evm/blob/a876368769ba01cea34271420c10a16862041f44/eth/abc.py#L634

I'm going to update this PR with a new `persist_unexecuted_block` to see what that feels like.

Reply:
I updated the PR. What do you think about this solution?